src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+		int32_t tmp = atoi(env_low_target);
+		if (tmp > 0) {
+			low_target = tmp;
+		}
+	}
+	int32_t high_target = 2 * low_target;
+	char* env_high_target = getenv("LIBDISPATCH_THREADPOOL_HIGH_TARGET");
+	if (unlikely(env_high_target)) {
+		int32_t tmp = atoi(env_high_target);
+		if (tmp >= low_target) {
+			high_target = tmp;
+		}
+	}
+	int32_t max = MIN(8 * high_target, 512);
+	char* env_max = getenv("LIBDISPATCH_THREADPOOL_MAX");
+	if (unlikely(env_max)) {
+		int32_t tmp = atoi(env_max);
+		if (tmp > 0) {
+			max = tmp;
+			low_target = MIN(low_target, max);
+			high_target = MIN(high_target, max);
+		}
+	}
+	target_runnable_workers = low_target;
+	target_max_runnable_workers = high_target;
+	max_total_workers = max;
+
+	// Now that we have set threadpool parameters, initialize the management subsystem
+	_dispatch_workqueue_management_init();
+
+	// Must come after _dispatch_workqueue_management_init() sets threadpool_verbose
+	verbose_printf("normal threadpool worker targets: %d...%d with max of %d",
+			target_runnable_workers, target_max_runnable_workers, max_total_workers);
+
+	// overcommit pool
+	_dispatch_allocate_workqueues(overcommit_pool.queues, WORKQ_NUM_PRIORITIES, true);
+	r = pthread_mutex_init(&overcommit_pool.mutex, NULL);
+	(void)dispatch_assume_zero(r);
+
+	r = pthread_cond_init(&overcommit_pool.spare_workers, NULL);
+	(void)dispatch_assume_zero(r);
+
+	// normal pool
+	_dispatch_allocate_workqueues(normal_pool.queues, WORKQ_NUM_PRIORITIES, false);
+
+	// create initial set of normal workers
+	for (int i=0; i<target_runnable_workers; i++) {
+		_dispatch_spawn_thread(_dispatch_normal_worker_main);
+	}
+}
+
+
+int
+dispatch_workqueue_get_wq(dispatch_workqueue_t * workqp,
+		int priority, int overcommit)
+{
+	dispatch_once_f(&_dispatch_workqueue_init_once_pred, NULL,
+					&_dispatch_workqueue_init_once);
+
+	dispatch_assert(priority >= 0 && priority < WORKQ_NUM_PRIORITIES);
+	if (overcommit) {
+		*workqp = overcommit_pool.queues[priority];
+	} else {
+		*workqp = normal_pool.queues[priority];
+	}
+
+	return 0;
+}
+
+int
+dispatch_workqueue_additem_np(dispatch_workqueue_t workq,
+		void (*workitem_func)(void *), void * workitem_arg)
+{
+	dispatch_workitem_t wi = _dispatch_alloc_workitem(workitem_func, workitem_arg);
+	unsigned int wq_index_bit = (0x1 << workq->priority);
+
+	if (unlikely(workq->overcommit)) {
+		// overcommit pool uses trival concurrency control: all operations done holding pool mutex
+		pthread_mutex_lock(&overcommit_pool.mutex);
+		STAILQ_INSERT_TAIL(&workq->item_listhead, wi, item_entry);
+		overcommit_pool.mask |= wq_index_bit;
+		if (overcommit_pool.num_spares > 0) {
+			pthread_cond_signal(&overcommit_pool.spare_workers);
+			overcommit_pool.num_spares--;
+		} else {
+			_dispatch_spawn_thread(_dispatch_overcommit_worker_main);
+		}
+		pthread_mutex_unlock(&overcommit_pool.mutex);
+	} else {
+		// normal pool uses finer-grained wq locking and atomic memory ops.
+		pthread_spin_lock(&workq->sl);
```

Comment:
please don't sue pthread_spin_lock() there are perfectly fine implementations of a better lock in dispatch itself (os/lock.h the unfair lock guy) you should use that instead.